### PR TITLE
Time dependent Navier Stokes

### PR DESCRIPTION
**Pull request by saidctb**
The following changes have been made:
- Accelerate the transpose method with different backends
- Add `minres`  and `lsmr` iterative solvers
- Fix a bug in the generated code of the dot product
- Add bent pipe geometry file
- Add a time-dependent Navier Stokes example
- Add `animate_field` method to visualize a field across time
- Add the option of putting the identity in the boundary element of a Matrix to make it invertable
- Add   `dtype` property to the abstract interface of the `VectorSpace`
- Add  `toarray` method to the abtract interface of the `Vector` class

### CODE REVIEW

**Comment by saidctb** [comment]
This PR is blocked by the Sympde PR [95](https://github.com/pyccel/sympde/pull/95)

File: psydac/linalg/iterative_solvers.py [hunk 1]
```diff
@@ -470,8 +474,8 @@ def bicg(A, At, b, x0=None, tol=1e-6, maxiter=1000, verbose=False):
         #-----------------------
         # MATRIX-VECTOR PRODUCTS
         #-----------------------
-        A .dot(p , out=v )
-        At.dot(ps, out=vs)
+        v  = A .dot(p )
+        vs = At.dot(ps)
```
Comment (yguclu) [inline psydac/linalg/iterative_solvers.py:478]: Why this change? It creates two temporary vectors for each iteration...

File: psydac/linalg/iterative_solvers.py [hunk 2]
```diff
@@ -521,3 +525,503 @@ def bicg(A, At, b, x0=None, tol=1e-6, maxiter=1000, verbose=False):
...
+    allowed: m = n, m > n, or m < n. ``b`` is a vector of length m.
+    The matrix A may be dense or sparse (usually sparse).
+
+
+    Parameters
+    ----------
+    A : psydac.linalg.basic.LinearOperator
+        Left-hand-side matrix A of linear system; individual entries A[i,j]
+        can't be accessed, but A has 'shape' attribute and provides 'dot(p)'
+        function (i.e. matrix-vector product A*p).
```
Comment (yguclu) [inline psydac/linalg/iterative_solvers.py:759]: The `At` parameter is missing, see docstring of `bicg` function.

**Review by yguclu — CHANGES_REQUESTED** [review]

File: psydac/api/ast/parser.py [hunk 3]
```diff
@@ -1003,14 +1003,21 @@ def _visit_ComputeKernelExpr(self, expr, op=None, lhs=None, **kwargs):
         target         = self._target
         dim            = self._dim
 
+        if normal_vectors:
+            axis    = target.axis
+            ext     = target.ext if isinstance(target, Boundary) else 1
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/ast/parser.py:1008]: this is the old implementation if will be fixed in the PR #107

File: psydac/api/ast/parser.py [hunk 4]
```diff
@@ -1003,14 +1003,21 @@ def _visit_ComputeKernelExpr(self, expr, op=None, lhs=None, **kwargs):
...
+
+        vars_plus = []
         if isinstance(target, Interface):
-            mapping = mapping.minus
+            target    = target.minus
+            mapping   = mapping.minus
+            # This is a hack, it must be deleted after merging PR #107
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/ast/parser.py:1014]: lol, this one si important , otherwise we will get the wrong results

File: psydac/linalg/tests/test_bicg.py [hunk 5]
```diff
@@ -23,6 +23,7 @@ def test_bicg_tridiagonal( n ):
     # Build generic non-singular matrix
     # Here tridiagonal matrix with values [-1, 3, -2] on diagonals
     A = np.diag([-1.0]*(n-1),-1) + np.diag([3.0]*n,0) + np.diag([-2.0]*(n-1),1)
+    A = np.diag([-1.0]*(n-1),-1) + np.diag([2.0]*n,0) + np.diag([-1.0]*(n-1),1)
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/linalg/tests/test_bicg.py:26]: it was a mistake

File: psydac/linalg/block.py [hunk 6]
```diff
@@ -88,6 +90,10 @@ def ends( self ):
     def pads( self ):
         return self._spaces[0].pads
 
+    @property
+    def dtype( self ):
+        return self._dtype
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/linalg/block.py:95]: the dtype here is the same as the dtype in `StencilVector`

File: psydac/api/essential_bc.py [hunk 7]
```diff
@@ -78,17 +79,20 @@ def apply_essential_bc_stencil(a, *, axis, ext, order):
         s = V.starts[axis]
         index = [(s + order if j == axis else slice(None)) for j in range(n)]
         a[tuple(index)] = 0.0
+        if isinstance(a, StencilMatrix) and identity:
+            a[tuple(index[:n//2])+(0,)*(n//2)] = 1.
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/essential_bc.py:83]: you don't need to , if you delete it by doing `b = b- A.dot(x0)`

File: psydac/api/ast/linalg.py [hunk 8]
```diff
@@ -291,6 +298,114 @@ def _compile_pyccel(self, mod, backend, verbose=False):
...
+
+        return folder
+
+    def _generate_code(self, backend=None):
+        dec  = ''
+        code = self._code
+        tag = random_string( 8 )
+        if backend and backend['name'] == 'pyccel':
+            imports = 'from pyccel.decorators import types'
+            dec     = '@types({})'.format(','.join(self._args_dtype))
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/ast/linalg.py:362]: the problem is with the old pyccel versions

[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/ast/linalg.py:362]: Oh yes, you're right. The minimum required Pyccel version is `0.10.1` and that did not support the type annotations yet. Could you please add a comment next to that line? We may want to remove it in the future.

File: psydac/api/ast/parser.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/ast/parser.py:1008]: I see, thanks.

File: psydac/api/essential_bc.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/essential_bc.py:83]: Before doing `b -= A.dot(x)`, was `b` already set to zero and `A` set to the identity at the boundary?

Does `x0` contain the coefficients of the lifting, which are zero inside the domain and non-zero on the boundary?

If that is so, for the boundary coefficients what you wrote corresponds to
```python
x = A^(-1) * (b - A * x0) = I * (0 - I * x0) = -x0
```
and so there is a wrong sign there...

File: psydac/api/ast/linalg.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/ast/linalg.py:362]: okay

File: psydac/api/essential_bc.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/essential_bc.py:83]: you do that before setting anything to zero , so your equation will be` A^(-1) * (b - A * x0) = I * (x - I * x0) = x-x0`

[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/essential_bc.py:83]: I see, so the order of operations is the following

```python
# 1. compute the lifting x0
x0 = ...

# 2. subtract lifting contribution from right-hand side
b -= A.dot(x0)

# 3. set boundary coefficients of A and b to zero, then set diag(A) = 1 on the boundary
apply_essential_bc(b, **kwargs)
apply_essential_bc(A, **kwargs, identity=True)

# 4. compute homogeneous solution x using some linear solver
x = cg(A, b)

# 5. add lifting to obtain non-homogeneous solution
x += x0
```
Is that what you are doing?

[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/essential_bc.py:83]: yes, in my case x0=0, so I didn't do that, but in general, it will be like this

[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/essential_bc.py:83]: Well, if `x0 = 0` you only need steps 3 and 4... 😄

File: psydac/linalg/block.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/linalg/block.py:95]: That does not answer my questions.

[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/linalg/block.py:95]: the dtype property, specifies the data types of the elements.
and yes I will add it to the abstract interface

File: psydac/linalg/iterative_solvers.py (hunk 1 above)
Comment (saidctb) [reply psydac/linalg/iterative_solvers.py:478]: yes I forgot to turn them back

File: psydac/api/fem.py [hunk 9]
```diff
@@ -789,6 +793,21 @@ def grid(self):
...
 
+    def get_space_indices_from_target(self, domain, target):
+        if domain.mapping:
+            domain = domain.logical_domain
+        if target.mapping:
+            target = target.logical_domain
+
+        domains = domain.interior.args
+        if isinstance(target, Interface):
+            raise NotImplementedError("TODO")
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/fem.py:804]: done

File: psydac/api/essential_bc.py [hunk 10]
```diff
@@ -41,7 +42,7 @@ def check_boundary_type(bc):
                 .format(type(bc)))
 
 #==============================================================================
-def apply_essential_bc_stencil(a, *, axis, ext, order):
+def apply_essential_bc_stencil(a, *, axis, ext, order, identity=False):
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/essential_bc.py:45]: done

File: psydac/linalg/stencil.py [hunk 11]
```diff
@@ -395,6 +395,10 @@ def ends(self):
     def pads(self):
         return self._space.pads
 
+    @property
+    def dtype( self ):
+        return self.space.dtype
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/linalg/stencil.py]: The dtype property specifies the data types of the elements of the StencilVector.
It's been added to the abstract interface

File: psydac/linalg/stencil.py [hunk 12]
```diff
@@ -1405,15 +1380,119 @@ def _update_ghost_regions_serial( self, direction: int ):
...
+                self._tranpose_args['ndT{i}'.format(i=i+1)] = ndiagsT[i]
+
+            for i in range(len(nrows)):
+                self._tranpose_args['si{i}'.format(i=i+1)] = si[i]
+
+            for i in range(len(nrows)):
+                self._tranpose_args['sk{i}'.format(i=i+1)] = sk[i]
+
+            for i in range(len(nrows)):
+                self._tranpose_args['sl{i}'.format(i=i+1)] = sl[i]
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/linalg/stencil.py]: done, I have merged all the loops into one loop

File: psydac/linalg/stencil.py [hunk 13]
```diff
@@ -1405,15 +1380,119 @@ def _update_ghost_regions_serial( self, direction: int ):
...
+
+    # ...
     def set_backend(self, backend):
-        from psydac.api.ast.linalg import LinearOperatorDot
-        self._backend = backend
-        self._args = self._dotargs_null.copy()
+        from psydac.api.ast.linalg import LinearOperatorDot, TransposeOperator
+        self._backend        = backend
+        self._args           = self._dotargs_null.copy()
+        self._tranpose_args  = self._tranpose_args_null.copy()
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/linalg/stencil.py]: fixed

File: psydac/linalg/tests/test_bicg.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/linalg/tests/test_bicg.py]: fixed

File: psydac/linalg/utilities.py [hunk 14]
```diff
@@ -52,3 +53,33 @@ def petsc_to_stencil(x, Xh):
     u = array_to_stencil(x, Xh)
     return u
 
+def _sym_ortho(a, b):
+    """
+    Stable implementation of Givens rotation.
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/linalg/utilities.py:58]: fixed

File: psydac/utilities/utils.py [hunk 15]
```diff
@@ -4,7 +4,7 @@
 
 import numpy as np
 
-#===============================================================================
+#============================================================================================================================
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/utilities/utils.py]: fixed

File: psydac/utilities/utils.py [hunk 16]
```diff
@@ -14,7 +14,7 @@ def refine_array_1d( x, n ):
 
     return xr
 
-#===============================================================================
+#============================================================================================================================
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/utilities/utils.py]: fixed

File: psydac/utilities/utils.py [hunk 17]
```diff
@@ -34,3 +34,42 @@ def unroll_edges(domain, xgrid):
 
     elif xgrid[-1] != xB:
         return np.array([*xgrid, xgrid[0] + (xB-xA)])
+
+#============================================================================================================================
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/utilities/utils.py]: fixed

File: psydac/linalg/block.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/linalg/block.py]: fixed

File: psydac/linalg/iterative_solvers.py (hunk 1 above)
Comment (saidctb) [reply psydac/linalg/iterative_solvers.py]: fixed

File: psydac/linalg/iterative_solvers.py (hunk 2 above)
Comment (saidctb) [reply psydac/linalg/iterative_solvers.py:758]: fixed

File: psydac/linalg/stencil.py [hunk 18]
```diff
@@ -1448,50 +1450,26 @@ def set_backend(self, backend):
...
-                self._tranpose_args['si{i}'.format(i=i+1)] = si[i]
-
-            for i in range(len(nrows)):
-                self._tranpose_args['sk{i}'.format(i=i+1)] = sk[i]
-
-            for i in range(len(nrows)):
-                self._tranpose_args['sl{i}'.format(i=i+1)] = sl[i]
+            for arg_name, arg_val in zip(args_names, args_vals):
+                for i in range(len(nrows)):
+                    self._transpose_args[arg_name.format(i=i+1)] = arg_val[i]
```
Comment (yguclu) [inline psydac/linalg/stencil.py]: Nice! I suggest that you use a dictionary instead of two lists, as it is much easier to modify it later without making a mistake. You can use the contents of `args_names` as keys, as they are all different strings.

File: psydac/api/essential_bc.py [hunk 19]
```diff
@@ -43,6 +43,28 @@ def check_boundary_type(bc):
...
 def apply_essential_bc_stencil(a, *, axis, ext, order, identity=False):
+    """ This function applies the homogeneous boundary condition to the Stencil objects,
+        by setting the boundary degrees of freedom to zero in the StencilVector,
+        and the corresponding rows in the StencilMatrix/StencilInterfaceMatrix to zeros.
+        If the identity keyword argument is set to True, the boundary diagonal terms are set to 1.
+
+    Parameters
+    ----------
+    a : LinearOperator/Vector
+        The LinearOperator/Vector that will be modified.
```
Comment (yguclu) [inline psydac/api/essential_bc.py:54]: Thanks for adding this nice docstring! I think we need to be a bit more specific here: the only types that we support are `StencilVector`, `StencilMatrix` and `StencilInterfaceMatrix`.

File: psydac/api/essential_bc.py [hunk 20]
```diff
@@ -43,6 +43,28 @@ def check_boundary_type(bc):
...
+
+    axis : int
+        Axis of the boundary.
+
+    ext : int
+        Extremity of the boundary.
```
Comment (yguclu) [inline psydac/api/essential_bc.py:60]: It would be useful to say that the only values allowed are -1 and 1.

File: psydac/api/essential_bc.py [hunk 21]
```diff
@@ -43,6 +43,28 @@ def check_boundary_type(bc):
...
+
+    order : int
+        order of the boundary condition.
```
Comment (yguclu) [inline psydac/api/essential_bc.py:63]: Here we should mention what we mean by `order`, and that it must be `>= 0`:
```python
"""
    order : int
        All function derivatives up to `order` are set to zero
        on the specified boundary. `order >= 0` is required.
"""
```

File: psydac/api/essential_bc.py [hunk 22]
```diff
@@ -43,6 +43,28 @@ def check_boundary_type(bc):
...
+
+    identity : bool
+        If true, set the diagonal terms to 1
```
Comment (yguclu) [inline psydac/api/essential_bc.py:66]: Here I would also be a bit more specific:
```python
"""
    identity : bool
        If true, the diagonal terms corresponding to boundary coefficients are set to 1.
"""
```

File: psydac/linalg/basic.py [hunk 23]
```diff
@@ -52,6 +59,11 @@ def shape( self ):
     def space( self ):
         pass
 
+    @property
+    @abstractmethod
+    def dtype( self ):
+        pass
+
```
Comment (yguclu) [inline psydac/linalg/basic.py:66]: My original comment was

> What is this `dtype` property?
> Why is it not part of the abstract interface of the `VectorSpace` base class?

So I was expecting you to answer my questions in writing, explaining why you need that property and where it is used. I did not say "Please add `dtype` to `VectorSpace", did I?

So I think that we should discuss about this.

cc: @campospinto

File: psydac/linalg/basic.py [hunk 24]
```diff
@@ -128,6 +140,11 @@ def domain( self ):
     def codomain( self ):
         pass
 
+    @property
+    @abstractmethod
+    def dtype( self ):
+        pass
+
```
Comment (yguclu) [inline psydac/linalg/basic.py:147]: See my previous comment.

I am not sure that `LinearOperator` should have the `dtype` method... As far as I can tell only the `Matrix` class strictly needs `dtype`, because it has the `toarray` and `tosparse` methods.

**Review by yguclu — CHANGES_REQUESTED** [review]

File: psydac/linalg/basic.py (hunk 23 above)
Comment (saidctb) [reply psydac/linalg/basic.py:66]: okay we can discuss about it

File: psydac/linalg/basic.py (hunk 24 above)
Comment (saidctb) [reply psydac/linalg/basic.py:147]: we need to know the datatypes of the elements that it accepts, whether it works on integers, complex numbers of floats

File: psydac/linalg/stencil.py (hunk 18 above)
Comment (saidctb) [reply psydac/linalg/stencil.py]: done

File: psydac/api/essential_bc.py (hunk 19 above)
Comment (saidctb) [reply psydac/api/essential_bc.py]: done

File: psydac/api/essential_bc.py (hunk 20 above)
Comment (saidctb) [reply psydac/api/essential_bc.py]: done

File: psydac/api/essential_bc.py (hunk 22 above)
Comment (saidctb) [reply psydac/api/essential_bc.py]: done

**Review by yguclu — CHANGES_REQUESTED** [review]
Can you please add the latest changes in the PR description?
Modifications to the abstract classes are very important!

**Review by yguclu — APPROVED** [review]
